### PR TITLE
[pipes] add report_component_exception

### PR DESCRIPTION
**Pull request by alangenfeld**
Establish a pattern with message reader threads to report exceptions to the handler. 

## How I Tested These Changes

manually tested with the s3 message reader and not having aws auth. Previously it would fail silently but now reports as:

Screenshot 2023-10-04 at 5 14 37 PM

will add test coverage if we like this approach

### CODE REVIEW

**Comment by alangenfeld** [comment]
Current dependencies on/for this PR:
* master
  * **PR #17006** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/17006" target="_blank">Graphite</a> 
    * **PR #17014** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/17014" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/17014?utm_source=stack-comment).

**Comment by alangenfeld** [comment]
some context: exceptions in threads get handled by the global `threading.excepthook` which default it just prints to stderr. Since were library code I don't think we can muck with that global. 

I tried a bit to package up the try except pattern and wasn't happy with what i came up with so just kept it direct.

File: python_modules/dagster/dagster/_core/pipes/context.py [hunk 1]
```diff
@@ -191,6 +194,16 @@ def _handle_log(self, message: str, level: str = "info") -> None:
...
 
+    def report_component_exception(self, origin: str, exc_info: ExceptionInfo):
+        # report as an engine event to provide structured exception data
+        DagsterEvent.engine_event(
+            self._context.get_step_execution_context(),
+            f"[pipes] exception occurred in component {origin}",
+            EngineEventData(
+                error=serializable_error_info_from_exc_info(exc_info),
+            ),
+        )
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/pipes/context.py:205]: i tried to change the other "system" pipes message to engine event to align this, but the front-end handling for system events doesn't handle warning and only errors if you have the exception payload. I did a quick stab at updating the front-end and engine event code to support "warnings" but it still looked scuffed so bailed back to this.

File: python_modules/dagster/dagster/_core/pipes/client.py [hunk 2]
```diff
@@ -73,6 +73,14 @@ def inject_context(self, context_data: "PipesContextData") -> Iterator[PipesPara
             process to locate and load the injected context data.
         """
 
+    @abstractmethod
+    def no_messages_debug_text(self) -> str:
+        """A message to be displayed when no messages are received from the external process to aid with debugging.
+
+        Example: "Attempted to inject context using a magic portal. Expected PipesMagicPortalContextLoader to be
```
[the remark this answers is not in the document]
Comment (schrockn) [reply python_modules/dagster/dagster/_core/pipes/client.py:80]: Can we provide a default implementation of `no_messages_debug_text`. I got bit by that I think Pipes customizer will as well.

File: python_modules/dagster/dagster/_core/pipes/context.py [hunk 3]
```diff
@@ -164,6 +194,16 @@ def _handle_log(self, message: str, level: str = "info") -> None:
         check.str_param(message, "message")
         self._context.log.log(level, message)
 
+    def report_component_exception(self, origin: str, exc_info: ExceptionInfo):
+        # report as an engine event to provide structured exception data
+        DagsterEvent.engine_event(
```
[the remark this answers is not in the document]
Comment (schrockn) [reply python_modules/dagster/dagster/_core/pipes/context.py:199]: Oh I see just the wording of the static method name here. I might the one named this function a long time ago but at this point the name is confusing and it should have a verb in it. Not your problem just opining.

File: python_modules/dagster/dagster/_core/pipes/utils.py [hunk 4]
```diff
@@ -268,20 +290,29 @@ def download_messages_chunk(self, index: int, params: PipesParams) -> Optional[s
...
+                            message = json.loads(line)
+                            handler.handle_message(message)
+                        self.counter += 1
+                    elif is_task_complete.is_set():
+                        break
+                time.sleep(1)
+        except:
+            handler.report_component_exception(
+                f"{self.__class__.__name__} reader thread",
+                sys.exc_info(),
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/pipes/utils.py:313]: Can we expand on error string here?

Exception thrown in the thread spawned by <<class nanme>> 

or something like that?

**Review by schrockn — CHANGES_REQUESTED** [review]
Req changes to discuss new term "component" in the pipes ontology

File: python_modules/dagster/dagster/_core/pipes/context.py [hunk 5]
```diff
@@ -164,6 +194,16 @@ def _handle_log(self, message: str, level: str = "info") -> None:
...
+            self._context.get_step_execution_context(),
+            f"[pipes] exception occurred in component {origin}",
+            EngineEventData(
+                error=serializable_error_info_from_exc_info(exc_info),
+            ),
+        )
```
[the remark this answers is not in the document]
Comment (alangenfeld) [reply python_modules/dagster/dagster/_core/pipes/context.py:205]: i think component feels intuitive but i think this all works about just as well if you remove `component`

File: python_modules/dagster/dagster/_core/pipes/context.py [hunk 6]
```diff
@@ -191,6 +194,16 @@ def _handle_log(self, message: str, level: str = "info") -> None:
         check.str_param(message, "message")
         self._context.log.log(level, message)
 
+    def report_component_exception(self, origin: str, exc_info: ExceptionInfo):
+        # report as an engine event to provide structured exception data
+        DagsterEvent.engine_event(
+            self._context.get_step_execution_context(),
+            f"[pipes] exception occurred in component {origin}",
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/pipes/context.py:201]: [1]

File: python_modules/dagster/dagster/_core/pipes/utils.py (hunk 4 above)
Comment (alangenfeld) [reply python_modules/dagster/dagster/_core/pipes/utils.py:313]: [1] is the pre-amble

**Comment by schrockn** [comment]
prefer `report_pipes_framework_exception` or something to the component languge

**Review by schrockn — DISMISSED** [review]
Ok. One last thing. Can you document what the error messages look like? In particular in the screenshot in the summary it looks like the type of the original exception is omitted, which makes the error confusing. Can we make sure to include enough context so things are actionable etc. Original stack trace, exception name. etc.

This would also be nice fodder for an explicit metadata message where we could attach nice error stack information.

**Comment by alangenfeld** [comment]
>  Can you document what the error messages look like? 
In the code or the PR? 

>  in the screenshot in the summary it looks like the type of the original exception is omitted, which makes the error confusing

I am not sure what you mean, the event in the screenshot has 
* in black the message we prefix with about the context of the exception 
* the first line in red the exception class and exception message 
* the foliowing lines in red the stack trace

**Comment by schrockn** [comment]
Whoops I confused this and the warning message below the error.

**Review by schrockn — APPROVED** [review]

**Pull request merged**